framework/master-key-provider-keyring.md

```diff
@@ -0,0 +1,125 @@
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Inputs
+
+On keyring initialization, a keyring MUST define the following:
+
+- [Master Key Provider](#master-key-provider)
+
+### Master Key Provider
+
+A [master key provider](master-key-provider-interface.md) that can decrypt data keys using available [master keys](master-key-interface.md)
+ as well as provide [master keys](master-key-interface.md) for use in encryption.
+
+## Operation
+
+### OnEncrypt
+
+This keyring MUST acquire a list of [master keys](master-key-interface.md) by calling the master key provider's [Get Master Keys for Encryption](master-key-provider-interface.md#get-master-keys-for-encryption)
+using the [encryption context](structures.md#encryption-context) from the [encryption materials](structures.md#encryption-materials) as input. If the 
+master key provider does not provide any master keys, OnEncrypt MUST fail.
+
+If the input [encryption materials](structures.md#encryption-materials) do not contain a [plaintext data key](structures.md#plaintext-data-key), OnEncrypt MUST
+attempt to generate a new plaintext data key and encrypt that data key by calling [Generate Data Key](master-key-interface.md#generate-data-key)
+on the first (primary) [master key](master-key-interface.md) provided by the master key provider.
+
+If the call to Generate Data Key does not succeed, OnEncrypt MUST NOT modify the encryption materials and MUST fail.
+
+If the call succeeds, OnEncrypt MUST do the following with the response from [Generate Data Key](master-key-interface.md#generate-data-key)
+- set the plaintext data key on the [encryption materials](structures.md#encryption-materials)
+- append a new [encrypted data key](structures.md#encrypted-data-key) to the encrypted data key list
+  in the [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - the [ciphertext](structures.md#ciphertext) is the response encrypted data key.
+  - the [key provider id](structures.md#key-provider-id) is the provider id of the primary master key.
+  - the [key provider information](structures.md#key-provider-information) is the key id of the primary master key.
+- append a new record to the [keyring trace](structures.md#keyring-trace) 
+  in the input [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - The string field KeyNamespace contains the provider id of the primary master key.
+  - The string field KeyName contains the key id of the primary master key.
+  - The [flags](structures.md#flags) field of this record includes the following flags:
+    - [GENERATED DATA KEY](structures.md#flags)
+    - [ENCRYPTED DATA KEY](structures.md#flags)
+    - [SIGNED ENCRYPTION CONTEXT](structures.md#flags) (if the master key supports encryption context signing)
```

Comment:
How do I know if it does or does not?

Reply:
I don't think there is any specified way to determine that, other than knowing that KMS and Raw AES support it. I could change it to say (if the master key is known to support encryption context signing, such as the KMS Master Key and Raw AES Master Key)